Fixed release process for the nodified UI

Mingling94 opened the pull request:
Will probably just use https://github.com/maxogden/electron-packager in
the future as release.sh reinvents the wheel. It's extra work to use
release.sh and we don't even utilize electron's cross-platform
capabilities as well (i.e. our executable doesn't have an icon as well
as other features)

lukechampine commented:
Looks great! I'm testing it now...need to announce the linode as a host first though (yeesh).

The placeholder text doesn't get cleared when you type/paste text, you just type right over it.

lukechampine commented on .gitignore:
```diff
@@ -5,3 +5,12 @@ node_modules
+renter
+sia.wallet
+sia.wallet.backup
+wallet
```
Sia shouldn't be running from this directory; there should be a `Sia/` or `siad/` folder in the `Sia-UI` directory that contains these.

lukechampine commented on lib/daemon.js:
```diff
@@ -42,7 +42,7 @@ module.exports = {
 
         // TODO: Allow arguments from config to daemon
-        daemonProcess = child_process.spawn(config.siad_cmd, {cwd: path.join(path.dirname(__dirname), "Sia")});
+        daemonProcess = child_process.spawn(config.siad_cmd, {cwd: path.join(path.dirname(__dirname))});
```
oh wait I take it back. But why move it to top-level?

seveibar replied:
hmm, I just assumed it would be better to have the siad command path from the config file- i agree that it'd be better to have the files contained within the Sia directory- I think I'll revert this

lukechampine commented on main.js:
```diff
@@ -103,4 +106,28 @@ function setupIPCHandlers() {
+            "title": "Ascii Code"
+        });
+
+        popupWindow.loadUrl('file://' + __dirname + '/lib/popups/asciicode.html');
```
this file isn't in the repo

lukechampine commented on main.js:
```diff
@@ -103,4 +106,28 @@ function setupIPCHandlers() {
+        });
+    });
+    ipc.on("close-popup", function(event){
+        console.log("clse-popup");
```
typo

lukechampine commented:
awesome, awesome. Popup text is a bit light but no one should be reading it anyway.

how tough would it be to add a "Copy text" button? If it's just a few lines we should add it here but if it's more involved I'll merge this and we can do it next PR.

seveibar commented:
boom

lukechampine commented:
Thanks! I think we're just about ready for release now.

Mingling94 committed:
Fixed release process for the nodified UI

Will probably just use https://github.com/maxogden/electron-packager in
the future as release.sh reinvents the wheel. It's extra work to use
release.sh and we don't even utilize electron's cross-platform
capabilities as well (i.e. our executable doesn't have an icon as well
as other features)

bitspill commented:
Once electron-packager has been integrated electron-builder can be used to create installers to make a more streamline system https://github.com/loopline-systems/electron-builder

Mingling94 commented:
Yush :D

Mingling94 commented:
@lukechampine

lukechampine commented on release.sh:
```diff
@@ -48,9 +48,9 @@ package() {
-		cp -R ../app/ "$ui/Sia.app/Contents/Resources/app/"
-		mkdir -p "$ui/Sia.app/Contents/Resources/app/Sia"
+		rm -rf "$ui/Sia.app/Contents/Resources/default_app"
+		rsync -av --progress ../ "$ui/Sia.app/Contents/Resources/app/" --exclude "release/" --exclude ".git/"
```
what is the reason for switching from `rm && cp` to `rsync`?

Mingling94 replied:
I only added the -f flag to rm. This shouldn't change anything

rsync allows excluding folders, such as `release` and `.git`

lukechampine replied:
hmm. Seems dangerous to be copying blindly from the top-level dir. I usually have a few untracked files lying around there. I guess this is related to the comment below.

DavidVorick replied:
I am also generally against copying the top level folder. You want to be explicit as possible so that no cruft makes it into the release, we've had that problem before.

DavidVorick replied:
also, is the `-f` flag necessary? None of the files in that folder should be protected, and if they are the developer should probably be warned before they are removed

lukechampine commented on release.sh:
```diff
@@ -48,9 +48,9 @@ package() {
-		mkdir -p "$ui/Sia.app/Contents/Resources/app/Sia"
+		rm -rf "$ui/Sia.app/Contents/Resources/default_app"
+		rsync -av --progress ../ "$ui/Sia.app/Contents/Resources/app/" --exclude "release/" --exclude ".git/"
+		mkdir -p "$ui/Sia.app/Contents/Resources/app/app/Sia"
```
`app/app`??

Mingling94 replied:
So here's the thing, if we're to have one package.json and be a more traditional electron folder structure, we put ALL of our source code in electron's folder structure. That folder is called `app`. We also have a folder called `app` that siad (inside `Sia` folder) goes into.

I could rename the `app` folder in our repo to something else like `ui` to eliminate this redundancy.

I was thinking about this and the most elegant solution might be to make a clear separation in the UI where the root level of the repo is for purely development's sake. This development shell would have it's own package.json with all the devDependencies that we currently have (jshint, electron-prebuilt, etc.). The `app` folder would then contain a second package.json that only has the node packages that are dependencies for the app to run (bignumber.js and jquery).

Using `electron-packager` will also present the same problem.

lukechampine replied:
That seems sensible to me. We should look at other electron projects and perhaps copy their approach.

Mingling94 replied:
Noted, though, try this out and execute building a release and see if it poses any problems. For now, master's release.sh doesn't work so this is a short-term solution.

DavidVorick replied:
app/app is definitely not okay. app/ui is also not okay, why do we need this double-layering when before we didn't?

This is the reason that I removed electron-packager before, I couldn't figure out how to not have it do a double-app folder in the release. Changing the name is just masking the problem, we want as simple of a directory structure as possible.

lukechampine replied:
seems like a good example (using electron packager): https://github.com/colonizers/colonizers-desktop

DavidVorick replied:
> I was thinking about this and the most elegant solution might be to make a clear separation in the UI where the root level of the repo is for purely development's sake. This development shell would have it's own package.json with all the devDependencies that we currently have (jshint, electron-prebuilt, etc.). The app folder would then contain a second package.json that only has the node packages that are dependencies for the app to run (bignumber.js and jquery).

Isn't that how it already works?

Mingling94 replied:
@DavidVorick Other electron apps **don't** have all their js, css, etc. in another folder. We do because separating out all the dot-anything files like .jshintrc and .jsdoc plus our documentation from the functional stuff is just neat and convenient.

Changing the names isn't masking the problem, there's just not a problem that the two `app` folders didn't provoke you to imagine. Having two package.jsons is highly unconventional and much larger of a problem. Maintaining an in-house developed wheel reinvention of a packager script is also unconventional and a problem.

Checkout #115, it makes use of electron-packager and is quite a deal less code that has to be version-bumped, debugged, etc.

DavidVorick commented:
merge conflicts?

Mingling94 commented:
This is being closed in favor of #115
